service-proxy.js

```diff
@@ -595,35 +604,33 @@ ServiceDispatchHandler.prototype.getPartialRange =
 function getPartialRange(serviceName, reason) {
     var self = this;
 
-    var range = self.computePartialRange(serviceName);
-    if (!range.isValid()) {
+    var partialRange = self.partialRanges[serviceName];
+    if (!partialRange) {
+        var exitNodes = self.egressNodes.exitsFor(serviceName);
+        var serviceChannel = self.getOrCreateServiceChannel(serviceName);
+        var relays = Object.keys(exitNodes).sort();
+        var workers = serviceChannel.peers.keys().sort();
+        partialRange = new PartialRange(
+            self.channel.hostPort,
+            self.minPeersPerWorker,
+            self.minPeersPerRelay
+        );
+        partialRange.compute(relays, workers);
+    }
+
+    if (!partialRange.isValid()) {
         // This should only occur if an advertisement loses the race with a
         // relay ring membership change.
         self.logger.warn('Relay could not find itself in the affinity set for service', self.extendLogInfo({
             serviceName: serviceName,
             reason: reason,
-            partialRange: range
+            partialRange: partialRange
```

Comment:
Dear reviewers:
- I could add the `.extendLogInfo` idiom to partialRange and use it here, thoughts?